api/genotypes_loader.api.inc

```diff
@@ -48,41 +48,59 @@ function genotypes_loader_get_postgresql_version() {
  */
 
  /**
-  *  Give the different errors that are possible separate error codes. Also allow a parameter to be given to
+  *  @TODO: Give the different errors that are possible separate error codes. Also allow a parameter to be given to
   *  either "handle errors" or not- this way, if we see a certain error (such as not unique), we can handle it
   *  from outside of the function.
   *
-  *  Also - don't forget to refer to Lacey's new helper function on Github. This function needs to evolve
-  *  regardless! Whether or not we go with copying from a file remains to be seen...
   */
 
 function genotypes_loader_helper_add_record_with_mode($record_type, $table, $mode, $select_values, $insert_values = array()) {
+
+  // Set some variables to abstract mode
+  $select_only = 0;
+  $insert_only = 1;
+  $both = 2;
+
   // the name of the primary key.
   $pkey = $table . '_id';
+
   // First we select the record to see if it already exists.
   $record = chado_select_record($table, array($pkey), $select_values);
-  // If we want to insert values, we can merge our values to have all the information we need
-  $values = array_merge($select_values, $insert_values);
+
   // If it exists and the mode is 1 (Insert Only) then return an error to drush.
   if (sizeof($record) == 1) {
-    if ($mode == 1) {
-      return drush_set_error(dt('Record "@record_type" already exists but you chose to only insert (mode=@mode). Values: ', array('@record_type' => $record_type, '@mode' => $mode)).print_r($values,TRUE));
+    if ($mode == $insert_only) {
+      return drush_set_error(dt('Record "@record_type" already exists but you chose to only insert (mode=@mode). Values: ', array('@record_type' => $record_type, '@mode' => $mode)).print_r($select_values,TRUE));
     }
     // Otherwise the mode allows select so return the value of the primary key.
     else {
       return $record[0]->{$pkey};
     }
   }
+
   // If more then one result is returned then this is NOT UNIQUE and we should report an
-  // error to the user -not just run with the first one.
+  // error to the user - not just run with the first one.
   elseif (sizeof($record) > 1) {
-    return drush_set_error(dt('Record "@record_type" is not unique. (mode=@mode). Values: ', array('@record_type' => $record_type, '@mode' => $mode)).print_r($values,TRUE));
+    return drush_set_error(dt('Record "@record_type" is not unique. (mode=@mode). Values: ', array('@record_type' => $record_type, '@mode' => $mode)).print_r($select_values,TRUE));
   }
+
   // If there is no pre-existing sample but we've been given permission to create it,
-  // then we will insert it :-).
-  elseif ($mode) {
+  // then insert it
+  elseif ($mode == $both) {
+
+    // If we want to insert values, we can merge our values to have all the information we need
+    $values = array_merge($select_values, $insert_values);
```

Comment:
So $insert_values doesn't stand-alone? This should be documented in the function docblock.

Reply:
Addressed in commit https://github.com/UofS-Pulse-Binfo/genotypes_loader/commit/1277bb2ef06c2db55a61777187950a4e50ff4e10 :+1: